### ISSUE TITLE
[Long-term] The idea of rewriting

### ISSUE THREAD

**Issue by raitonoberu** [alexmercerind/youtube-search-python#104]
While working on adding a Hashtag feature, I carefully studied the code. Currently, the code is divided into the sync version and the async version, and a lot of code is copied instead of being reused. This complicates the maintenance & testing. In addition, the response data is stored in the search class (in 3 different variables), which significantly increases the use of RAM.

I'm sure that you guys have also developed as programmers since the first parts of this code were written, and now you can see a lot of bad practices that you haven't seen before. New features continue to be added, but the current library structure is simply not scalable - all new classes are added to the file `extra.py`, which is already difficult to read.

That's why we need to rewrite the project from scratch, creating a new, scalable and efficient structure. As for backward compatibility, I think it's better to provide instructions for migrating to a future version, because there are critical problems that cannot be fixed in any other way.

What I want to **implement** in the future:
- Scalable structure. Each feature will be stored in a separate file.
- Code reuse for the asynchronous version.
- A single code style.
- Using httpx for both versions. We're installing it anyway, so why not?
- A single API for both versions - `search.next()`.
- Data models instead of dictionaries, but it should be possible to use them as dictionaries.
- A way to combine a filter & a sort order.
- Automatic tests.

What I want to **get rid of** in the future:
- Storing data in a class. It is the user's problem what to do with the returned data.
- Making an HTTP request in the constructor. Declaration of variables, nothing more.
- Heavy inheritance. We are not in Java to write tons of classes in which we use only one method.

I know that you are no longer actively developing this project and are now focused on other projects, so I don't want you to worry. **I will be working on my own fork** and will make a pull request later. Now I want to discuss the idea of rewriting and breaking backward compatibility for the sake of a bright future 🌅

Thanks.

**Comment by alexmercerind** [comment]
Initially when I started working on this, it was just a single function library that only used to search for videos.
Later on, tons and tons of other classes, Video, Playlist, whole async version came in & I just slapped everything on community requests.
As time went by, that `extras.py` got filled with nearly "primary" features instead of "extras" (this can definitely be improved).

Idea of rewriting can be good, but if you notice very closely, then you'll find out that this library has never-ever made any breaking changes. There are still many libraries/bots/scripts that depend on this & changing API will cause them to break. e.g. even old 'SearchVideos` etc. etc. is still present, when it has been replaced by `VideosSearch` now.
As a reason, I introduced async version into whole new subpackage `__future__`. Similar for other classes, instead of changing functionality of a feature (like changing to async, changing class behaviour etc.), I preferred to introduce new one. As a project, we now have over 8.1k dependent repositories (and that just open source), it lies as a responsibility to keep the API consistent/non-changing.

The idea of code sharing between async/sync might seem good on paper, but then you'll notice that you have to specifically mark "async def" for async functions & "def" for sync functions (so, both had to be mutually exhaustive). And, at the end you will end up making code structure more confusing in my opinion while sharing function calls etc. The idea of inheritance lies because it is not good idea to expose internal methods/variables in user accessible class.

**Comment by alexmercerind** [comment]
Anyways, I do agree that code is not good at this point & can definitely be improved. I don't mind adding you as collaborator if you wish.

**Comment by raitonoberu** [comment]
> The idea of code sharing between async/sync might seem good on paper, but then you'll notice that you have to specifically mark "async def" for async functions & "def" for sync functions (so, both had to be mutually exhaustive)

You are mistaken. I'll try to explain it now, but I advise you to read somewhere about asynchronous Python.
<details>

Functions can be blocking, or non-blocking. If the function is blocking, the execution of the program stops until the function is completed.
```python
# Here is an example of a blocking function. It does stop the execution.
def getSyncResponse(url):
    return urlopen(url)  # do some IO-related job

# Here is an example of a non-blocking function. It does not stop the execution.
def find_min(a, b):
    if a > b:
        return a
    return b
```
Asynchronous Python uses only one thread. The "await" keyword tells the event loop to go and do something else while this async function is executing. **If there is no "await" keyword in the function, there is no need to mark it as "async"**.
```python
# Asynchronous function
async def getAsyncResponse(url):
    # we can call other async functions
    a = await session.get(url)  # some IO-related job is async now

    # we can call non-blocking function form asynchronous code
    find_min(a.status_code, 200)

    # we can NOT call blocking function form asynchronous code
    # getResponse(url)

    return a.json()
```
Let's go back to the real code.
https://github.com/alexmercerind/youtube-search-python/blob/544d16888fb261dd92e0db98d22edf6aac532b14/youtubesearchpython/__future__/handlers/componenthandler.py#L74-L89
This function is non-blocking. In fact, you just marked the sync function as async. I do not know if this affects the performance. Instead, you can simply reuse already written functions.

In addition, this entire file is a mistake:
https://github.com/alexmercerind/youtube-search-python/blob/544d16888fb261dd92e0db98d22edf6aac532b14/youtubesearchpython/__future__/internal/json.py#L1-L12
`json.loads`/`json.dumps` are **non-blocking** function (unlike in JS, i know). httpx itself just calls json.loads when you call `response.json()`. Running an executor is an expensive operation, it may even be a bottleneck.

Thank you for reading this, I hope it has become clearer to you now.
</details>

**Comment by raitonoberu** [comment]
> As a project, we now have over 8.1k dependent repositories (and that just open source), it lies as a responsibility to keep the API consistent/non-changing.

That is why `setup.py` provides the ability to include a specific version of the library:
```python
setup(
    ...,
    install_requires=[
        'youtube-search-python==1.4.6',
    ],
)
```
And it is the developer's responsibility to update dependencies and code. I don't know...

**Comment by alexmercerind** [comment]
So, will adding this "common code" & boilerplate "improve" the scalability? 
That's why I decided to keep both completely separate & apart from one another.

A computation is a computation, there is no nature of being "blocking" or "non blocking". And being just an IO does not specify if it is blocking or not. A single loop indefinitely will be blocking (even if not indefinitely). An IO operation can be non-blocking in other words. There is difference between thread idling & sleeping, I don't know what you are calling as "blocking".
Parsing a JSON is a CPU intensive task & in dynamically typed language like Python, one can just deserialize JSON into a map, which means there is no specification how much long or how many loops will take place (or how the JSON is going to be).
A discord bot for example, most likely to play music will definitely be performing some kind of event-polling in parallel, thus being non-blocking is necessary for this library. Infact, this whole async `__future__` subpackage was part of a issue (closed now), saying that discord.py is async & this library should be too. So maybe you wanna give some resources about "json.loads/json.dumps are non-blocking function"?

> This function is non-blocking. In fact, you just marked the sync function as async. I do not know if this affects the performance. Instead, you can simply reuse already written functions.

There needs to be clear speculation between what you are considering "blocking" or "non-blocking".

Python's "async programming" is a real joke. No big project even uses it, and neither performance is main goal of Python. As I said, many discord bots use this library & there was an issue regarding async version, so made it because of discord.py being async in nature.

> That is why setup.py provides the ability to include a specific version of the library.

How do I tell you that, not-just consistent API, but **being functional** is also part of this library. Imagine **someone using an old version & suddenly it stops working because YouTube changed something**, and now one has to upgrade regardless & face the API changes anyway.

**Comment by alexmercerind** [comment]
What can certainly be improved is to have a common interface for all requests. But, even that won't work because some use the new innertube and some depend on POST `pbj=1` and other request kinds... This library is just collection of everything I figured out as time went by.

And I certainly admit things can be done better, because I was also learning when I made this & still don't call myself expert in any way about Python.

**Comment by mytja** [comment]
I will do parts of this next week. If anyone wants to do something, you can assign yourself. I updated original comment to contain Tasklist & what I will do.
Thanks

**Comment by mytja** [comment]
This is how I imagined code reuse: https://github.com/mytja/youtube-search-python/commit/d3582462571d144916128bdcd27422112e0255e4
I got rid of VideoInternal class, because now, we will use a single class, that comes with async functions aswell as sync functions. This is very basic for now, and I will certanly improve maintainability of Core class, but this is just a first look.
I also implemented automated testing using GitHub Actions: https://github.com/mytja/youtube-search-python/runs/3644078464?check_suite_focus=true
As you can see, now on one codebase, sync Video (which worked before) works, and also async Video (which didn't work before) works.

**Comment by mytja** [comment]
So, I migrated every single Internal class in extras to Core classes & now everything that wasn't working before works. Classes in search.py are going to be fairly easy to migrate.

But problem is, and I don't know if it was before - when I place Hashtag search (asynchronous) at the top of function, it works very well, but if I place it somewhere more deeper in function (see asyncExample.py), it will fail. This is definetly strange.

**Comment by raitonoberu** [comment]
I see only one possible reason:
https://github.com/mytja/youtube-search-python/blob/2e739ef71b2c31e8d40a9f7b867933e5747a0b0a/youtubesearchpython/core/hashtag.py#L81-L87
The body was "polluted" by other requests, because you are passing the same map everywhere. Consider replacing such lines with `copy.deepcopy(requestPayload)` everywhere in the code. I'm sure it will help! 😉
